Now accepts non-ambiguous partial long options,  a la GNU getopt_long…

Bill-Gray opened the pull request:
First: I added the Github URL for your project to the header so I'll know whence the code came and where to look for updates.

Second:  GNU getopts_long() allows truncating long options as long as the result isn't ambiguous.  If you have options --color and --collapse,  for example,  --colo or --colla or --coll will be accepted.  --co or --col won't be.  I've revised `optparse` to behave similarly.

The GNU version does give a better error message;  for example, --co would get you

`./getopt: option '--co' is ambiguous; possibilities: '--color' '--collapse'`

I haven't gotten quite that fancy.  (If I did,  `msgbuf[64]` would probably need enlarging.)

Third and not least... thank you very much!  I have code that has to work on Linux,  *BSD,  and Windows; other getopts wouldn't have worked for me.

Bill-Gray committed:
Now accepts non-ambiguous partial long options,  a la GNU getopt_long.  Added reference to original version.

Bill-Gray committed:
Changed many functions and values to be of type const,  where possible.  (I usually declare the argv array as 'const char **argv'.)  One can still pass a non-const argv,  using a cast.

skeeto committed:
Add optind changes by seishinryohosha: In case of a missing option argument,  this change prevents optind from wandering off the end of the argument list should option parsing attempt to continue.

Bill-Gray committed:
README should mention a core strength of Optparse,  i.e.,  you can use it on non-POSIX systems (getopt is unreliable in that regard).

Krasjet committed:
Make makefile POSIX compatible

For POSIX make, the `$<` macro is only defined for inference rules and
the `.DEFAULT` rule. It is unspecified in target rules, e.g. for BSD
make it is left empty, so `make run` will fail under BSDs.

skeeto committed:
Support shortname values outside of ASCII (#16)

This is more in line with GNU getopt_long(), and it's more flexible.

Bill-Gray committed:
Revised 'test' to use const variables where possible

Bill-Gray committed:
Improved alignment for a (slightly) more compact structure.  From E5ten/optparse@b6524d0d221f3a93b7291.

skeeto committed:
Add examples/, including subcommand example (#18)

skeeto committed:
Fix subcommands example (no "p" switch)

ShamrockLee commented:
You can `git rebase`-ing your commits onto master instead of cherry-picking commits from the master branch to keep the history clean.

Bill-Gray commented:
Hmmm... thank you;  that's some food for thought.

My use of Git has been somewhat basic.  I assume you're thinking along the following lines.  With my changes (d, e, f),  we had

```
skeeto   a - b - c
                  \
Bill-Gray          d - e - f
```
Then @skeeto added three commits (call them g, h, i),  which I cherry-picked.  He then added some more :
```
skeeto   a - b - c - g - h - i - plus_more
                  \
Bill-Gray          d - e - f - g - h - i
```
Ideally,  I'd drop the cherry-picks,  rebase and shift to...
```
skeeto   a - b - c - g - h - i - plus_more
                                      \
Bill-Gray                              d - e - f
```
...and,  as skeeto added further commits,  shift along to keep up.

I kinda like this idea.

I am assuming none of my changes are going to be pulled (or used by anybody except me,  it appears),  so that part is not much of a concern.

However,  it'd be slightly cleaner.  I haven't really been all that bothered by this,  but I have another situation where this sort of history-mangling is more annoying.  It may make sense for me to experiment with git-rebase in this (probably simple) case,  then try it on the more annoying (and much more complicated) situation.

Alternatively,  I could simply delete everything,  re-fork the original,  and apply my three patches.  That would be simpler and not require me to learn anything,  but wouldn't help me learn what I might need to know to tackle the complicated situation.

N-R-K committed:
don't assume non-zero argc

the C standard doesn't prohibit a zero argc, from C89 5.1.2.2:

> * The value of argc shall be nonnegative.
> * argv[argc] shall be a null pointer.
> * If the value of argc is greater than zero, the string pointed to by
>   argv[0] represents the program name (...)

this situation can be triggered by calling exec with NULL argv[0] and
tested by applying the following patch to test.c:

	     if (argc > 1) {
	+        execv(argv[0], (char *[]){NULL});
	+    } else if (argc == 0) {
	         return manual_test(argc, argv);
	     } else {
	         return testsuite();

not sure how likely it is for such situation to occurs in practice,
but deal with it in any case by setting optind to 0 if argv[0] happens
to be null.

Bill-Gray committed:
Include an actual test suite in test.c (from skeeto/optparse@f0c65c9,  modified slightly)

Bill-Gray committed:
Revised 'examples' for correct use of const.

Bill-Gray committed:
README.md : added a paragraph to describe differences between this and the parent repository.

ShamrockLee commented:
@Bill-Gray One uses

```
git fetch -p <upstream remote>
git rebase --onto=<upstream remote>/<base branch> HEAD~<number of commits to move>
```

to keep their feature branch up with the upstream, and

```
git push -f <fork remote>
```

to update their PR.

It's often a bad idea to merge the base branch update into the feature branch the PR, since it brings unrelated commits and file changes into the current PR, and brings repetitive commit into the source tree after merge. GitHub even provides a repository configuration option that blocks PR with non-linear history from merging.

Bill-Gray commented:
@ShamrockLee - (groan) I'd completely forgotten that this pull request even existed.  It's not been pulled after six years,  and presumably won't ever be,  and the history has gotten hopelessly scrambled.  I'll just close this out.
